### ISSUE TITLE
discussion: Future of WyldCard

### ISSUE THREAD

**Issue by ooper-zz** [defano/wyldcard#98]

**Comment by ooper-zz** [comment]
WyldCard has unequivocally achieved the original academic goals, and perhaps has lent itself for some personal goals and challenges. So Kudos for that and for the outstanding architecture that came out of it. But it begs the question:  What unresolved problems can WyldCard solve today, and for who?

**Comment by ooper-zz** [comment]
No thoughts on this yet. Perhaps I can start with some open, more focused Q&A as food-for-thought might help start a discussion toward what the future holds for WyldCard: 

For starters, let us define success as "mass adoption" for now.

What made HyperCard successful/failure?
A. HyperCard was successful when all Macs had it built-in and stacks were generally unlocked (ala Open Source) and free --although you had to buy it if your Mac didn't come with it.  The card concept ignited the interest of non-programmers as well as aspiring ones, and signed up some great integration partners, such as Oracle as clients to back ends and Kiosks of all sorts. In my opinion, HyperTalk was the secret sauce for non-programmers and aspiring ones alike. 

With that said, no color, Mac only, single user, no Internet integration, as well as inability to easily generate what most wanted at the time: double-clickable, standalone, document-based applications. A Claris takeover (a product subsidiary of Apple) tried to address some of these issues, but the price tag was too steep and breaking it up into authoring and runtime environments threw a monkey wrench to the vibrant community that was keeping it alive. Apple took it back while I was there, and at one point, it was targeted to adopt the QuickTime file format -- a scriptable QuickTime was a promising idea, even today. Yes, Jobs killed it, but as aspiring programmers had moved on and non-programmers didn't want to pay for something that was once free --it was already kinda dead. Enter the clones...

 What is it that no clones have really succeeded? 
A. A combination of factors.  They wanted to productize it, several broke up authoring from its runtime, applied untested business models, had no mass appeal, and quickly fell into feature creep --not to mention the advent of the Internet "browser", which had reached some maturity with Java. 

That brings us to WyldCard.  What can we learn from the above lessons to try to define what a successful adoption might look like?
A.  Non-programmers and programmers have increased exponentially, and open-source and programmers have also increased.  I think a hybrid 

Java does not play well with mobile apps. Do we care?

**Comment by defano** [comment]
Thanks for starting this conversation, @ooper-zz. I have several thoughts on this:

First, if I were going to write WyldCard from scratch today, I’d do it in Dart and Flutter where it could be cross-compiled for iOS, Android, macOS, Windows, Linux and the web. Java is "cross platform" but (realistically) only cross desktop-platforms.  

I think the magic of HyperCard was that it allowed a 12-year-old kid (me) to write application software that was state-of-the-art. You could make a HyperCard stack back in the early nineties that was every bit as sophisticated and professional as commercial software written in Pascal or C++. That was amazing!

A successful, modern successor would make it possible for lay people to take advantage of contemporary computing technologies like the the cloud, AI, augmented reality, etc. I suspect that no modern equivalent exists today because the nature of modern "apps" is fundamentally more complicated than it was back then. For example:

1. **Cross-platform, responsive user interfaces**: HyperCard, you’ll recall, had fixed window (stack) dimensions. This was acceptable in 1989, but not today. Building a user interface that can resize itself and even reorganize itself depending on available real estate remains astoundingly complicated, even for folks like us that do this professionally. Maybe someday an AI-based solution will exist to auto-magically rejigger UI widgets based on available space, but today this remains the hardest part of UI development.  Plus, applications aren’t limited to software just running on your desktop anymore. Users expect desktop and mobile variants of the same app with a cloud-hosted backend to share data between them all. 
2. **The cloud**: A modern HyperCard would have its own cloud, in some form or another. But once you introduce a cloud-hosted backend you’re stuck dealing with information assurance; account and credential management; etc. I’m sure there are clever ways to simplify those problems, but even then you’d be left to solve the economic one: Servers in the cloud cost money. A kid playing around with this would incur fees in perpetuity. Ick. Advertisements? Double-ick. 
3. **Social networking**: I am biased towards younger audiences when thinking about what a modern HyperCard should be (because of my own experience with it). And in this regard, I would think a contemporary version of HyperCard would have to be capable of building social applications that make it easy for users to share just about anything with just about anybody. I suppose if you can solve the distributed systems problem, this one might not be too far behind.  

As for WyldCard, well, there's still some low-hanging fruit we ought to consider:

* We really, really, _really_, need to spend a bit of time on dev-ops/CI. We ought to be generating distributable packages with every commit and making them readily available for users who are not Java programmers and don't have ready access to Gradle. 
* We ought to add support for parts of the language that were implemented as in-built XFCNs and XCMDs... things like the palette editor, picture viewer, sound recorder, colorizeHC, etc. This will improve backwards compatibility.  
* Then we ought to invent our own XCMD/XFCN concept that would allow other Java programmers to extend WyldCard just the way HyperCard's externals did. This should be pretty simple. 
* User levels, report printing, stack script inheritance and some other language areas need to be implemented. There are a couple visual effects that I didn't implement, either. 
* There's an issue opened already, but the MaryTTS (voice) library appears to be poorly maintained, at best. Out of scope for this project, but it sure would be nice to have a more modern text-to-speech framework. At the very least, perhaps somebody could help assure those Mary libraries are readily published to a well-maintained Maven repo.
* A lot more testing—especially with HyperCard stacks—to shake out bugs and incompatibilities.
* Performance is probably "good enough" (especially when considering how stacks performed on 90s-era hardware), but it's pretty terrible by today's standards.
* ... and if somebody was so graphically inclined, it would be nice to have a professionally-designed launcher icon for the app, too. And maybe some better documentation.

**Comment by ooper-zz** [comment]
Thank you for the insights, @defano!
It’s Intriguing to hear your personal background with HyperCard :). It was and is amazing...it inspired so many of us and will always be special. I have a six-year-old grandchild and at Apple I helped develop a Kid simulation platform called Cocoa (predates the framework with the same name), so I hear you about the young audience :); although I think you would agree that we need to suitable for all ages: "Simply, Powerful", as we called our products at Claris (before I moved to Apple). Even though it totally screwed up HyperCard, LOL.

**Reactions**

_on the future_:
On the UI I have nothing to add.  Funny you mentioned Dart/Flutter, I was hoping to bring it up :). I can't think of a better platform, especially given that Java and Dart are so similar. 

We can do it in stages. First, Antlr 4.9 now targets Dart, so we can leverage com.defano.hypertalk in WyldCard-F. Similarly, we can also leverage com.defano.wyldcard.part components. I think that we can learn a great deal about the user’s needs once we release WyldCard-J. Moreover, if we embed a browser in WyldCard-J, we can start rolling out new functionality on the cloud while still being able to run desktop stacks, which I think is still needed in one form or another. At the very least WyldCard-J ought to be able to optionally host its stacks as multi-user resources with card-level locking and no Level 4 (authoring). Not to stray too far, but I think p2p deployment deserves some thought as well and possibly a major market leadership and certainly a way to avoid the costs in cloud servers.

As far as extensions, Yes XCMD/XFCNs need to be supported, and we ought to be able to introduce the notion of “objects” in the language as well.  Yes, some missing features and I think authoring needs to be at least as good as HyperCard's. This is to compensate any perceived modern-UI shortcomings. I’ve been testing by importing stacks with more to go; some important stuff is still buggy or missing.  
The HyperCard community is still very much alive and they are definitely free press, if we do this well! For docs, we can start by pointing to Danny Goodman’s books, since we are the only HyperCard compatible to-date, that's a plus. I will bring this up to him to see if his books are still printed or printable. I suspect that if we do well on imports, help will come on the writing and graphic needs. 

Good stuff! Thanks for the insights again.

**Comment by ooper-zz** [comment]
**lifeline for today's WyldCard**

The biggest strength in HyperCard wasn't so much the development of shrink-wrap applications of its time. Yes, many of us would have wanted it, but in reality, it could never match the memory, performance, and UI-rich requirements. By far, the greatest majority of stack developers programmed for themselves (or for others), and it was always with a single user in mind and it was always local and personal: utilities, small databases, games, prototyping and the like. I believe the notion of personal scriptable buttons, fields, cards, backgrounds plus stacks, as well as local, personal "apps" is still needed today.  Moreover, while I believe that the need for multi-user and mobile "apps" is already fulfilled, there is no reason WyldCard can't interoperate and still thrive. 

So, instead of competing with the dozens of today's rich, drag-n-drop App-development tools, WyldCard can co-exist by being the glue that ties it all up. A scriptable Web component would place WyldCard in a different category, more as a competitor to the Web Browser itself --a hybrid App/Web browser, if you will. To get it done, we can leverage JavaFX WebView as it's based on WebKit. Whether we drop Swing for JavaFX (already 14 years old) is a whole other question, but perhaps there is a better path forward. I am hoping that we can leverage Web View as a path forward to something like Flutter.  

In summary, I propose we add the above feature and release WyldCard as is ASAP. We'll get some press and developer interest, which will take us to the next level: fix some minor bugs and improve authoring, which in my opinion, comes short of HyperCard 2.2's authoring environment.

**Comment by ooper-zz** [comment]
@defano In the above pursuit (WebView feature), I went ahead and integrated a WebView as a "Page" WyldCard object that can be dropped unto a card or background. It was meant as feasibility study, and led to exploratory space where I can surmise the following possibilities:

- Pages tailored to multiple audiences, such as kids, older adults, special needs of either, or any other tailoring needed for groups to improve or better control their workflow, none of which a canned Internet Browser can accomplish easily. 
- With additional work, we could use the capability to create page channels. Ever wanted to add a comment to a news feed, or ANY newsfeed? Unlike any commonly used Internet Browser, now you would have native web-based social and browsing, but also remote AND local stacks, bundled in a personalizable, customizable hybrid solution. I will expand on this at a later point.

Functional workflow
- An author creates a "Page" object, using "New Page" under the "Objects" menu. S/he further views and modifies it using its "Page Info..." menu item; such things as the Home web page for the object. 
- a "Page" object can then be owned by the card or background, with all the expected WyldCard editing functionality.  
- In HyperCard, any button is able to "LinkTo" cards or stacks. We can finally add it to WyldCard to include linking to a website.  Similarly to HyperCard, this would be done through a mouseUp script modification.

I am attaching a technical feasibility prototype video.

https://user-images.githubusercontent.com/3395566/193985062-5f79bf15-9852-4364-8883-beab28f41806.mov
